src/SegmentedMsg.chpl

```diff
@@ -12,10 +13,167 @@ module SegmentedMsg {
   use RandArray;
   use IO;
   use Map;
+  use GenSymIO;
 
   private config const logLevel = ServerConfig.logLevel;
   const smLogger = new Logger(logLevel);
 
+  /**
+  * Build a Segmented Array object based on the segments/values specified.
+  **/
+  proc assembleSegArrayMsg(cmd: string, payload: string, st: borrowed SymTab): MsgTuple throws {
+    var msgArgs = parseMessageArgs(payload, 2);
+    var segName = msgArgs.getValueOf("segments");
+    var valName = msgArgs.getValueOf("values"); 
+    smLogger.debug(getModuleName(),getRoutineName(),getLineNumber(),
+            "cmd: %s segmentsname: %s valuesName: %s".format(cmd, segName, valName));
+
+    var repMsg: string;
+    var segments = getGenericTypedArrayEntry(segName, st);
+    var segs = toSymEntry(segments, int);
+
+    var valEntry = st.tab.getBorrowed(valName);
+    if valEntry.isAssignableTo(SymbolEntryType.SegStringSymEntry){ //SegString
+      var vals = getSegString(valName, st);
+      var segArray = getSegArray(segs.a, vals.values.a, st);
+      repMsg = "created " + st.attrib(segArray.name);
+    } 
+    else { // pdarray
+      var values = getGenericTypedArrayEntry(valName, st);
+      select values.dtype {
+        when (DType.Int64) {
+          var vals = toSymEntry(values, int);
+          var segArray = getSegArray(segs.a, vals.a, st);
+          repMsg = "created " + st.attrib(segArray.name);
+        }
+        when (DType.UInt64) {
+          var vals = toSymEntry(values, uint);
+          var segArray = getSegArray(segs.a, vals.a, st);
+          repMsg = "created " + st.attrib(segArray.name);
+        }
+        when (DType.Float64) {
+          var vals = toSymEntry(values, real);
+          var segArray = getSegArray(segs.a, vals.a, st);
+          repMsg = "created " + st.attrib(segArray.name);
+        }
+        when (DType.Bool) {
+          var vals = toSymEntry(values, bool);
+          var segArray = getSegArray(segs.a, vals.a, st);
+          repMsg = "created " + st.attrib(segArray.name);
+        }
+        otherwise {
+            throw new owned ErrorWithContext("Values array has unsupported dtype %s".format(values.dtype:string),
+                                        getLineNumber(),
+                                        getRoutineName(),
+                                        getModuleName(),
+                                        "TypeError");
+        }
+      }
+    }
+    smLogger.debug(getModuleName(), getRoutineName(), getLineNumber(), repMsg);
+    return new MsgTuple(repMsg, MsgType.NORMAL);
+  }
+
+  /**
+  * Procedure to access the Segments of a Segmented Array
+  * This may not be needed once all functionality is server side
+  **/
+  proc getSASegmentsMsg(cmd: string, payload: string, st: borrowed SymTab): MsgTuple throws {
+    var repMsg: string = "";
```

Comment:
i hate how we gotta create a new `SymEntry` for `getSASegments/Values` when we already have one, we just don't have the name... but like you say, these functions won't be needed once the segarray functionality is severside

Reply:
I may have an idea to work around this in the meantime that would be useful in other areas as well, but wanted to get this up. I will look into implementation options and see if we can get around this, but for now we won't have a choice.